File: Part 2 - Regression/Section 6 - Polynomial Regression/Poly_regr.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
#library(caTools)
#set.seed(123)
#split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

#Fitting Linear Regression to the dataset
regr = lm(formula = Salary ~., data = dataset)



#Fitting Polynomial Regression to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
regr_poly = lm(formula = Salary ~ ., data = dataset)


#Visualizing Linear Regression
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(regr, newdata = dataset)),
            color = 'blue') +
  ggtitle('Truth or Bluff(Linear Regression)') +
  xlab('Level') +
  ylab('Salary')


#Visualizing Polynomial Regression

ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(regr_poly, newdata = dataset)),
            color = 'blue') +
  ggtitle('Truth or Bluff(Poly Regression)') +
  xlab('Level') +
  ylab('Salary')

#Predicting new result with Linear Regression
y_pred = predict(regr, data.frame(Level = 6.5))


#Predicting new result with Polynomial Regression
y_pred = predict(regr_poly, data.frame(Level = 6.5, 
                                       Level2 = 6.5^2,
                                       Level3 = 6.5^3,
                                       Level4 = 6.5^4))

File: Part 2 - Regression/Section 6 - Polynomial Regression/Poly_regr.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, -1].values

# Splitting the dataset is not needed due to small dataset
"""from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

#Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
regr = LinearRegression()
regr.fit(X, y)

#Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_regr = PolynomialFeatures(degree = 4)
X_poly = poly_regr.fit_transform(X)
regr2 = LinearRegression()
regr2.fit(X_poly, y)

#Visualize Linear results
plt.scatter(X, y, color = 'red')
plt.plot(X, regr.predict(X), color = 'blue')
plt.title('Truth or Bluff')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

#Visualize Poly Results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regr2.predict(poly_regr.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff')
plt.xlabel('Level')
plt.ylabel('Salary')
plt.show()

#Predict a new result with Linear Regression
regr.predict([[6.5]])

#Predict a new result with Polynomial Regression
regr2.predict(poly_regr.fit_transform([[6.5]]))






